File: src/log.c
/* vi: set cindent:
 * $Id$
 *
 * ROX-Session, a very simple session manager
 * Copyright (C) 2002, Thomas Leonard, <tal197@users.sourceforge.net>.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the Free
 * Software Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 * Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* log.c - handles the log window which captures standard error output */

#include "config.h"

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <fcntl.h>
#include <time.h>

#include <gtk/gtk.h>
#include <gdk/gdkx.h>

#include "global.h"

#include "main.h"
#include "session.h"
#include "log.h"
#include "options.h"
#include "gui_support.h"

#define MARGIN 4

static Option o_time_shown;
static Option o_percent_switch;

static GtkWidget *log_window = NULL;	/* The popup message window */
static GtkWidget *label = NULL;	/* Contains the log messages */
static gint	 input_tag;

static int	real_stderr = -1;

/* Concat all the chunks together to get the current message... */
typedef struct _Chunk Chunk;
struct _Chunk {
	time_t time;
	gchar  *text;
};
static GList	*chunks = NULL;

/* This holds history of recent log messages, which can be displayed
 * in a normal window.
 */
#define MAX_BUFFER_SIZE 100000			/* In characters */
static time_t last_log_time;			/* Time of timestamp */
static GtkTextBuffer *buffer = NULL;
static GtkWindow *message_window = NULL;	/* The non-popup window */

/* Static prototypes */
static void got_log_data(gpointer data,
			 gint source,
			 GdkInputCondition condition);
static gint log_clicked(GtkWidget *text, GdkEventButton *bev, gpointer data);
static void log_msg(const gchar *text, gint len);
static GList *show_log(Option *option, xmlNode *node, guchar *label);
void show_message_log(void);
static void log_own_errors(const gchar *log_domain,
			   GLogLevelFlags log_level,
			   const gchar *message,
			   gpointer user_data);

/****************************************************************
 *			EXTERNAL INTERFACE			*
 ****************************************************************/

void log_init(void)
{
	GtkTextIter end;
	int fds[2];

	buffer = gtk_text_buffer_new(NULL);
	gtk_text_buffer_create_tag(buffer,
				"time", "foreground", "blue",
				NULL);
	time(&last_log_time);
	gtk_text_buffer_get_end_iter(buffer, &end);
	gtk_text_buffer_insert(buffer, &end, "ROX-Session started: ", -1);
	gtk_text_buffer_insert_with_tags_by_name(buffer, &end,
				ctime(&last_log_time), -1,
				"time", NULL);

	option_add_int(&o_time_shown, "log_time_shown", 5);
	option_add_int(&o_percent_switch, "log_percent_switch", 30);

	option_register_widget("show-log", show_log);

	if (pipe(fds))
	{
		g_warning("pipe(): %s\n", g_strerror(errno));
		return;
	}

	/* Since writing to stderr will block until we're ready to
	 * read it, it's a good idea to get our own errors some other
	 * way (ROX-Session may deadlock due because it can't write the
	 * error until it's ready to read previous errors, and can't read
	 * previous errors until it's written the new one).
	 * Pango likes to spew errors if it can't find its fonts...
	 */
	g_log_set_handler("", G_LOG_LEVEL_WARNING | G_LOG_FLAG_FATAL
                     | G_LOG_FLAG_RECURSION, log_own_errors, NULL);
	g_log_set_handler("Gdk", G_LOG_LEVEL_WARNING | G_LOG_FLAG_FATAL
                     | G_LOG_FLAG_RECURSION, log_own_errors, NULL);
	g_log_set_handler("Gtk", G_LOG_LEVEL_WARNING | G_LOG_FLAG_FATAL
                     | G_LOG_FLAG_RECURSION, log_own_errors, NULL);

	/* Grab a copy of stderr before we replace it.
	 * We'll duplicate output here if it exists...
	 */
	real_stderr = dup(STDERR_FILENO);

	/* Dup the writeable FD to stderr */
	if (fds[1] != STDERR_FILENO)
	{
		if (dup2(fds[1], STDERR_FILENO) == -1)
		{
			g_warning("dup2(): %s\n", g_strerror(errno));
			return;
		}
		close(fds[1]);
	}

	/* Don't close on exec()... */
	if (fcntl(STDERR_FILENO, F_SETFD, FALSE))
		g_warning("fcntl(): %s\n", g_strerror(errno));

	/* But, do close the other end */
	if (fcntl(fds[0], F_SETFD, TRUE))
		g_warning("fcntl(): %s\n", g_strerror(errno));

	log_window = gtk_window_new(GTK_WINDOW_POPUP);
	gtk_window_set_resizable(GTK_WINDOW(log_window), FALSE);
	gtk_widget_set_name(log_window, "log_window");
	gtk_widget_realize(log_window);
	gtk_widget_add_events(log_window, GDK_BUTTON_PRESS_MASK);
	gtk_signal_connect(GTK_OBJECT(log_window), "button_press_event",
			GTK_SIGNAL_FUNC(log_clicked), NULL);

	label = gtk_label_new(NULL);
	gtk_misc_set_alignment(GTK_MISC(label), 0.0, 0.0);
	gtk_widget_add_events(label, GDK_BUTTON_PRESS_MASK);
	gtk_container_add(GTK_CONTAINER(log_window), label);
	gtk_widget_show(label);

	input_tag = gdk_input_add(fds[0], GDK_INPUT_READ, got_log_data, NULL);
}

/****************************************************************
 *			INTERNAL FUNCTIONS			*
 ****************************************************************/

/* Chunks has changed. Update the label (or hide if nothing to show).
 * If there's too much stuff, open the log window.
 */
static void show(void)
{
	static gboolean top = FALSE;
	int	y;
	GString *text;
	GList	*next;
	int	last_non_space = -1;
	GdkRectangle mon_geom;

	if (message_window || !chunks || !o_time_shown.int_value)
	{
		gtk_widget_hide(log_window);
		return;
	}

	/* Always use monitor 0, keeps it simpler */
	if (gdk_screen_get_n_monitors(gdk_screen_get_default()) > 1)
	{
		gdk_screen_get_monitor_geometry(gdk_screen_get_default(), 0,
				&mon_geom);
	}
	else
	{
		mon_geom.x = mon_geom.y = 0;
		mon_geom.width = gdk_screen_width();
		mon_geom.height = gdk_screen_height();
	}
	if (!GTK_WIDGET_VISIBLE(log_window))
	{
		gdk_window_get_pointer(GDK_ROOT_PARENT(), NULL, &y, NULL);

		top = 2 * y > mon_geom.height + mon_geom.y;
	}

	text = g_string_new(NULL);
	for (next = chunks; next; next = next->next)
	{
		Chunk   *chunk = (Chunk *) next->data;

		g_string_append(text, chunk->text);
	}

	{
		gboolean nl = TRUE;
		int i;

		for (i = 0; i < text->len; i++)
		{
			if (text->str[i] == '\n')
			{
				if (nl)
					g_string_erase(text, i, 1);
				nl = TRUE;
			}
			else
			{
				last_non_space = i;
				nl = FALSE;
			}
		}
	}

	g_string_truncate(text, last_non_space + 1);

	gtk_label_set_text(GTK_LABEL(label), text->str);

	g_string_free(text, TRUE);

	if (text->len)
	{
		GtkRequisition req;
		int	max_h;

		gtk_widget_size_request(label, &req);
		max_h = o_percent_switch.int_value * mon_geom.height / 100;

		if (req.height > max_h)
		{
			show_message_log();
			return;
		}

		if (top)
			y = mon_geom.y + MARGIN;
		else
			y = mon_geom.y + mon_geom.height - MARGIN - req.height;

		req.width = mon_geom.width - 2 * MARGIN;

		gtk_widget_set_uposition(log_window, mon_geom.x + MARGIN, y);
		gtk_widget_set_usize(log_window, req.width, req.height);
		gtk_widget_show(log_window);
		gdk_window_move(log_window->window, MARGIN, y);
		gdk_window_raise(log_window->window);
	}
	else
		gtk_widget_hide(log_window);
}

/* Remove old entries */
static gboolean prune(gpointer data)
{
	static  gint prune_timeout = 0;
	gboolean did_prune = FALSE;
	time_t	prune_time;

	prune_time = time(NULL) - o_time_shown.int_value;

	while (chunks)
	{
		Chunk *first = (Chunk *) chunks->data;

		if (first->time > prune_time)
			break;		/* Still want to show this */

		chunks = g_list_remove(chunks, first);
		g_free(first->text);
		g_free(first);

		did_prune = TRUE;
	}

	if (did_prune)
		show();

	if (chunks)
	{
		if (prune_timeout)
			gtk_timeout_remove(prune_timeout);
		prune_timeout = gtk_timeout_add(1000, prune, NULL);
	}

	return FALSE;
}

static void log_msg(const gchar *text, gint len)
{
	Chunk   *new;
	int	chars;
	GtkTextIter end;
	time_t  now;

	if (len < 0)
		len = strlen(text);

	time(&now);

	/* Add to the long history buffer... */
	gtk_text_buffer_get_end_iter(buffer, &end);
	if (now - last_log_time > 60)
	{
		gtk_text_buffer_insert_with_tags_by_name(buffer,
						 &end, ctime(&now), -1,
						 "time", NULL);
		last_log_time = now;
	}
	gtk_text_buffer_insert(buffer, &end, text, len);

	/* Remove stuff from the beginning if it gets too long... */
	chars = gtk_text_buffer_get_char_count(buffer);
	if (chars > MAX_BUFFER_SIZE)
	{
		GtkTextIter start;

		gtk_text_buffer_get_start_iter(buffer, &start);
		end = start;
		gtk_text_iter_forward_chars(&end, chars - MAX_BUFFER_SIZE);
		gtk_text_iter_forward_line(&end);
		gtk_text_buffer_delete(buffer, &start, &end);
	}


	/* And to the popup message area... */
	new = g_new(Chunk, 1);

	time(&new->time);
	new->text = g_strndup(text, len);
	chunks = g_list_append(chunks, new);

	prune(NULL);
	show();
}

static void log_own_errors(const gchar *log_domain,
			   GLogLevelFlags log_level,
			   const gchar *message,
			   gpointer user_data)
{
	g_print("%s\n", message);
}

static void write_stderr(guchar *buffer, int len)
{
	int	pos = 0, sent;

	while (pos < len)
	{
		sent = write(real_stderr, buffer + pos, len - pos);
		if (sent < 0)
		{
			log_msg(_("ROX-Session: write to stderr failed!\n"),
					-1);
			close(real_stderr);
			real_stderr = -1;
			return;
		}
		pos += sent;
	}
}

#define BUFFER_SIZE 256

static void got_log_data(gpointer data,
			 gint source,
			 GdkInputCondition condition)
{
	guchar	buffer[BUFFER_SIZE];
	int	got;

	got = read(source, buffer, BUFFER_SIZE);

	if (got <= 0)
	{
		gtk_input_remove(input_tag);
		log_msg(_("ROX-Session: read(stderr) failed!\n"), -1);
	}
	else
	{
		if (g_utf8_validate(buffer, got, NULL)) {
			log_msg(buffer, got);
		} else {
			char *fixed;

			fixed = g_convert_with_fallback(buffer, got,
					"utf-8", "iso-8859-1",
					"?", NULL, NULL, NULL);
			log_msg(fixed, -1);
			log_msg("(invalid UTF-8)\n", -1);
			g_free(fixed);
		}

		if (real_stderr != -1)
			write_stderr(buffer, got);
	}

	if (call_child_died)
		child_died_callback();
}

void show_message_log(void)
{
	GtkWidget *view, *dialog, *sw, *vbox, *hbox, *close_button;
	GdkRectangle mon_geom;

	if (message_window)
	{
		gtk_window_present(message_window);
		return;
	}

	/* Can't guarantee to open on same monitor as pointer, but most people
	 * will have monitors with matching resolutions */
	if (gdk_screen_get_n_monitors(gdk_screen_get_default()) > 1)
	{
		int mx, my;

		gdk_window_get_pointer(GDK_ROOT_PARENT(), &mx, &my, NULL);
		gdk_screen_get_monitor_geometry(gdk_screen_get_default(),
				gdk_screen_get_monitor_at_point(
					gdk_screen_get_default(), mx, my),
				&mon_geom);
	}
	else
	{
		mon_geom.x = mon_geom.y = 0;
		mon_geom.width = gdk_screen_width();
		mon_geom.height = gdk_screen_height();
	}

	dialog = gtk_window_new(GTK_WINDOW_TOPLEVEL);
	message_window = GTK_WINDOW(dialog);
	gtk_window_set_title(message_window, _("ROX-Session message log"));
	vbox = gtk_vbox_new(FALSE, 0);
	gtk_container_add(GTK_CONTAINER(dialog), vbox);
	hbox = gtk_hbox_new(FALSE, 0);
	gtk_box_pack_end(GTK_BOX(vbox), hbox, FALSE, FALSE, 8);
	close_button = gtk_button_new_from_stock(GTK_STOCK_CLOSE);
	gtk_box_pack_end(GTK_BOX(hbox), close_button, FALSE, FALSE, 18);
	g_signal_connect_swapped(close_button, "clicked",
			G_CALLBACK(gtk_widget_destroy), message_window);
	g_signal_connect(message_window, "destroy",
			G_CALLBACK(gtk_widget_destroyed), &message_window);

	view = gtk_text_view_new_with_buffer(buffer);
	gtk_text_view_set_editable(GTK_TEXT_VIEW(view), FALSE);
	gtk_text_view_set_cursor_visible(GTK_TEXT_VIEW(view), FALSE);
	gtk_text_view_set_wrap_mode(GTK_TEXT_VIEW(view), GTK_WRAP_WORD);
	gtk_widget_set_size_request(view, 400, 100);

	sw = gtk_scrolled_window_new(NULL, NULL);
	gtk_scrolled_window_set_policy(GTK_SCROLLED_WINDOW(sw),
	                               GTK_POLICY_AUTOMATIC,
	                               GTK_POLICY_AUTOMATIC);
	gtk_scrolled_window_set_shadow_type(GTK_SCROLLED_WINDOW(sw),
					GTK_SHADOW_IN);
	
	gtk_box_pack_start(GTK_BOX(vbox), sw, TRUE, TRUE, 0);
	gtk_container_add(GTK_CONTAINER(sw), view);

	gtk_window_set_default_size(message_window,
				    mon_geom.width / 2, mon_geom.height / 4);
	/* I thought 1/6 of screen height looked too small */

	gtk_widget_show_all(GTK_WIDGET(message_window));

	gtk_widget_hide(log_window);
}

static gint log_clicked(GtkWidget *text, GdkEventButton *bev, gpointer data)
{
	if (bev->button == 1)
		gtk_widget_hide(log_window);
	else
		show_message_log();

	return 1;
}

static GList *show_log(Option *option, xmlNode *node, guchar *label)
{
	GtkWidget *button, *align;

	g_return_val_if_fail(option == NULL, NULL);
	
	align = gtk_alignment_new(0.5, 0.5, 0, 0);
	button = button_new_mixed(GTK_STOCK_YES, _("Show message log"));
	gtk_container_add(GTK_CONTAINER(align), button);
	gtk_signal_connect(GTK_OBJECT(button), "clicked",
			GTK_SIGNAL_FUNC(show_message_log), NULL);

	return g_list_append(NULL, align);
}

File: log.py
from rox import g
from rox.options import Option
import time, os, sys, gobject
import fcntl
import codecs

import rox

from_utf8 = codecs.getdecoder('utf-8')
from_latin1 = codecs.getdecoder('iso-8859-1')

o_time_shown = Option('log_time_shown', 5)
o_percent_switch = Option('log_percent_switch', 30)

MARGIN = 4

real_stderr = None

MAX_LOG_SIZE = 100000	# If complete message log exceeds this many characters, trim it (100K)

def close_on_exec(fd, close):
	fcntl.fcntl(fd, fcntl.F_SETFD, close)

def init():
	global real_stderr, log
	log_r, log_w = os.pipe()

	# Since writing to stderr will block until we're ready to
	# read it, it's a good idea to get our own errors some other
	# way (ROX-Session may deadlock due because it can't write the
	# error until it's ready to read previous errors, and can't read
	# previous errors until it's written the new one).
	# Pango likes to spew errors if it can't find its fonts...

	#import warnings
	#def showwarning(message, category, filename, lineno, file = None,
	#		 showwarning = warnings.showwarning):
	#	print category, filename, lineno, message
	#warnings.showwarning = showwarning

	# Grab a copy of stderr before we replace it.
	# We'll duplicate output here if it exists...
	real_stderr = os.dup(2)

	sys.stderr = sys.stdout		# Send errors from us to stdout to avoid deadlock

	# Dup writable stream to stderr
	if log_w != 2:
		os.dup2(log_w, 2)
		os.close(log_w)
	
	close_on_exec(2, False)
	close_on_exec(log_r, True)

	log = Log()
	def got_log_data(src, cond):
		global real_stderr
		got = os.read(src, 1000)
		if not got:
			g.input_remove(input_tag)
			log.log('ROX-Session: read(stderr) failed!\n')
			return False
		try:
			log.log_raw(got)
			while got and real_stderr is not None:
				written = os.write(real_stderr, got)
				if written < 0:
					log.log('ROX-Session: write to stderr failed!')
					os.close(real_stderr)
					real_stderr = None
					break
				got = got[written:]
		except:
			rox.report_exception()
		return True
			
	#os.write(2, 'hi - ' + chr(0xef) + '\n')

	input_tag = g.input_add(log_r, g.gdk.INPUT_READ, got_log_data)

class Log(object):
	raw_input_buffer = ''
	buffer = None		# GtkTextBuffer
	popup = None
	log_window = None
	chunks = []
	chunks_cleanup = None
	last_timestamp_logged = None

	def __init__(self):
		self.buffer = g.TextBuffer()
		self.buffer.create_tag('time', foreground = 'blue')

		self.last_timestamp_logged = time.time()

		end = self.buffer.get_end_iter()
		self.buffer.insert(end, "ROX-Session started: ")
		now = time.ctime(self.last_timestamp_logged)
		self.buffer.insert_with_tags_by_name(end, now + '\n', "time")

	def log_raw(self, data):
		self.raw_input_buffer += data
		try:
			message = from_utf8(self.raw_input_buffer)[0]
			self.raw_input_buffer = ''
		except UnicodeDecodeError:
			# Not valid UTF-8. Maybe we're part way through a
			# multi-byte sequence. Just go up to newline.
			nl = self.raw_input_buffer.rfind('\n')
			if nl == -1:
				return	# No newline. Wait for more input.

			chunk = self.raw_input_buffer[:nl]
			self.raw_input_buffer = self.raw_input_buffer[nl + 1:]
			
			# Got a complete chunk of UTF-8 (no half characters)
			try:
				message = from_utf8(chunk)[0]
			except UnicodeDecodeError:
				# Still not valid. Try latin1, so we can print something.
				message = from_latin1(chunk, 'replace')[0] + ' (invalid UTF-8)'
		self.log(message)

	def log(self, message):
		# Remove blank lines and trailing spaces
		lines = [line.lstrip() for line in message.split('\n')]
		message = '\n'.join(filter(None, lines))

		end = self.buffer.get_end_iter()

		# Log the time, but only if it hasn't been too long (30
		# seconds) since the last timestamp.
		now = time.time()
		if now > self.last_timestamp_logged + 30:
			self.buffer.insert_with_tags_by_name(end, time.ctime(now) + '\n', 'time')
			self.last_timestamp_logged = now
		self.buffer.insert(end, message + '\n')
		
		self.prune()

		if self.log_window and self.log_window.flags() & g.VISIBLE:
			# Full log window already open
			self.show_log_window()
		else:
			# Otherwise try the popup
			self.chunks.append(Chunk(message, time.time()))
			self.schedule_chunks_cleanup()
			self.show_popup()
	
	def prune(self):
		"""If self.buffer is too long, remove lines from the start."""
		chars = self.buffer.get_char_count()
		if chars <= MAX_LOG_SIZE:
			return
		start = self.buffer.get_start_iter()
		end = self.buffer.get_start_iter()
		end.forward_chars(chars - MAX_LOG_SIZE)
		end.forward_line()		# Remove complete lines only
		self.buffer.delete(start, end)
	
	def show_popup(self):
		if self.popup is None:
			self.popup = Popup(self)
		self.popup.show()
	
	def show_log_window(self):
		if self.popup and self.popup.flags() & g.VISIBLE:
			self.popup.hide()
		if self.log_window is None:
			self.log_window = LogWindow(self.buffer)
		self.chunks = []
		self.log_window.show()
	
	def schedule_chunks_cleanup(self):
		if self.chunks_cleanup is not None:
			gobject.source_remove(self.chunks_cleanup)
			self.chunks_cleanup = None
		self.expire_chunks()
		if not self.chunks:
			return
		delay = self.chunks[0].timestamp + o_time_shown.int_value - time.time()
		def expire():
			self.chunks_cleanup = None
			self.schedule_chunks_cleanup()
			return False
		self.chunks_cleanup = gobject.timeout_add(int(max(delay * 1000, 0)) + 1, expire)
	
	def expire_chunks(self):
		earliest_to_keep = time.time() - o_time_shown.int_value
		changed = False
		while self.chunks and self.chunks[0].timestamp < earliest_to_keep:
			del self.chunks[0]
			changed = True
		if changed:
			if self.chunks:
				self.show_popup()
			elif self.popup:
				self.popup.hide()

class LogWindow(g.Dialog):
	buffer = None

	def __init__(self, buffer):
		g.Dialog.__init__(self)

		self.buffer = buffer
		self.set_title(_("ROX-Session message log"))
		#self.set_name('log_window')
		self.set_has_separator(False)

		self.add_button(g.STOCK_CLOSE, g.RESPONSE_OK)

		self.tv = g.TextView(buffer)
		self.tv.set_size_request(400, 100)
		self.tv.set_editable(False)
		self.tv.set_cursor_visible(False)

		swin = g.ScrolledWindow(None, None)
		swin.set_policy(g.POLICY_AUTOMATIC, g.POLICY_AUTOMATIC)
		swin.set_shadow_type(g.SHADOW_IN)
		swin.add(self.tv)

		swin.show_all()
		self.vbox.pack_start(swin, True, True, 0)

		screen = g.gdk.screen_get_default()
		mx, my, mask = g.gdk.get_default_root_window().get_pointer()
		monitor = screen.get_monitor_at_point(mx, my)
		geometry = screen.get_monitor_geometry(monitor)
		self.set_default_size(geometry.width / 2, geometry.height / 4)

		def resp(box, r):
			self.hide()
		self.connect('response', resp)

		def delete(box, dev):
			return True
		self.connect('delete-event', delete)
	
	def show(self):
		cursor = self.buffer.get_mark('insert')
		self.tv.scroll_to_mark(cursor, 0)
		g.Dialog.show(self)
	
class Popup(g.Window):
	def __init__(self, log):
		g.Window.__init__(self, g.WINDOW_POPUP)
		#self.set_resizable(False)
		self.set_name('log_window')
		self.realize()
		self.add_events(g.gdk.BUTTON_PRESS_MASK)
		self.connect('button-press-event', self.clicked)

		self.log = log
		self.label = g.Label(buffer)
		self.label.set_alignment(0.0, 0.0)
		self.add(self.label)
		#label.add_events(g.gdk.BUTTON_PRESS_MASK)
		self.label.show()
	
	def clicked(self, win, bev):
		self.hide()
		if bev.button != 1:
			self.log.show_log_window()
	
	def show(self):
		message = ''
		for c in self.log.chunks:
			message += c.message
		# Remove blank lines
		message = '\n'.join([m for m in message.split('\n') if m.strip()])
		self.label.set_text(message)

		screen = g.gdk.screen_get_default()
		mx, my, mask = g.gdk.get_default_root_window().get_pointer()
		monitor = screen.get_monitor_at_point(mx, my)
		geometry = screen.get_monitor_geometry(monitor)

		req_width, req_height = self.label.size_request()
		max_h = o_percent_switch.int_value * geometry.height / 100

		top = 2 * my > geometry.height + geometry.y

		if req_height > max_h:
			self.hide()
			self.log.show_log_window()
			return

		if top:
			y = geometry.y + MARGIN
		else:
			y = geometry.y + geometry.height - MARGIN - req_height

		req_width = geometry.width - 2 * MARGIN

		#self.set_uposition(geometry.x + MARGIN, y)
		self.set_size_request(req_width, req_height)
		g.Window.show(self)
		self.window.move_resize(MARGIN, y, req_width, req_height)
		self.window.raise_()

class Chunk:
	def __init__(self, message, timestamp):
		self.message = message
		self.timestamp = timestamp
